src/libexpbuf/doc/libexpbuf.3
.\" man page for libexpbuf
.\" Contact dev@hyper-active.com.au to correct errors or omissions. 
.TH libexpbuf 3 "2 February 2010" "1.2.30" "Library for Expanding Buffer."
.SH NAME
libexpbuf \- library for an expanding buffer
.SH SYNOPSIS
.\" Syntax goes here. 
.B #include <expbuf.h>
.sp
.B typedef struct
.br
.B {
.br
.B 	char *data;
.br
.B 	unsigned int length;
.br
.B 	unsigned int max;
.br
.B } expbuf_t;
.sp

.B expbuf_t * expbuf_init(expbuf_t *buf, unsigned int size)
.br
.B void expbuf_clear(expbuf_t *buf)
.br
.B expbuf_t * expbuf_free(expbuf_t *buf)
.br
.B void expbuf_add(expbuf_t *buf, void *data, unsigned int len)
.br
.B void expbuf_purge(expbuf_t *buf, unsigned int len)
.br
.B void expbuf_shrink(expbuf_t *buf, unsigned int extra)
.br
.B char * expbuf_string(expbuf_t *buf)
.sp
compile with the 
.B -lexpbuf
option
.SH EXAMPLES
#include <expbuf.h>
.sp
int main(void) {
.br
	expbuf_t *buffer;
.br
	char data[1024]];
.br
	int len, avail;
.sp
	buffer = expbuf_init(NULL, 0);
.br
	len = read_data(data, 1024);
.br
	while (len >= 0) {
.br
		expbuf_add(buffer, data, len);
.br
		processed = process_data(BUF_DATA(buffer), BUF_LENGTH(buffer));
.br
		expbuf_purge(buffer, processed);
.br
		len = read_data(data);
.br
	}
.br
	buffer = expbuf_free(buffer);
.br
	assert(buffer == NULL);
.sp
	return(0);
.br
}
.SH DESCRIPTION
.B libexpbuf
is a simple library for handling an expanding buffer.  This buffer is very primative, 
and is merely a single data string (keeping track of the actual length), and a maximum 
size that has so far been allocated.
.sp
.SH FILES
.P 
.I /usr/lib/libexpbuf.so
.br
.I /usr/include/expbuf.h
.SH SEE ALSO
.BR expbuf_add (3),
.BR expbuf_clear (3),
.BR expbuf_free (3),
.BR expbuf_init (3),
.BR expbuf_purge (3),
.BR expbuf_set (3),
.BR expbuf_shrink (3),
.BR expbuf_string (3),
.BR expbuf_t (3)
.SH BUGS
No known bugs at this time. 
.SH AUTHOR
.nf
Clint Webb (webb.clint@gmail.com)
on behalf of Hyper-Active Systems, Australia.
.br
http://www.hyper-active.com.au/libexpbuf/
.fi


src/libexpbuf/libexpbuf.c
/*

	libexpbuf
	(c) Copyright Hyper-Active Systems, Australia

	Contact:
		Clinton Webb
		webb.clint@gmail.com

*/


#include "expbuf.h"

#include <assert.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>


#if (EXPBUF_VERSION != 0x00010230)
#error "Incorrect header version.  code and header versions must match."
#endif




// initialise an expbuf structure, assuming that it contains garbage to begin with.
expbuf_t * expbuf_init(expbuf_t *b, unsigned int size)
{
	expbuf_t *buf;
	
	// if a NULL is passed in, that means that we need to create the object ourselves (the safest path)
	if (b) {
		buf = b;
		buf->internally_created = 0;
	}
	else {
		buf = malloc(sizeof(expbuf_t));
		buf->internally_created = 1;
	}
	
	BUF_LENGTH(buf) = 0;
	if (size > 0) {
		BUF_DATA(buf) = (char *) malloc(size);
		if (BUF_DATA(buf) == NULL) { BUF_MAX(buf) = 0; }
		else { BUF_MAX(buf) = size; }
	}
	else {
		BUF_DATA(buf) = NULL;
		BUF_MAX(buf) = 0;
	}
	
	return(buf);
}

// clear out any data in the buffer.  This will reset the length pointer only.  It will not free any resources held.
void expbuf_clear(expbuf_t *buf)
{
	assert(buf);
	assert(BUF_LENGTH(buf) <= BUF_MAX(buf));
	assert((BUF_DATA(buf) == NULL && BUF_LENGTH(buf) == 0 && BUF_MAX(buf) == 0) || (BUF_DATA(buf) != NULL && BUF_MAX(buf) > 0));
	
	BUF_LENGTH(buf) = 0;
}

// this will clear out any data in the buffer, and free the resources
// allocated.  After calling this function, it should be safe to
// deallocate the expbuf_t object.
//
// if the object was created internally, returns a NULL.  If the object was not 
// created internally, then it will return the object so that it can be deleted 
// elsewhere.
expbuf_t * expbuf_free(expbuf_t *buf)
{
	assert(buf);
	assert(BUF_LENGTH(buf) <= BUF_MAX(buf));
	assert((BUF_DATA(buf) == NULL && BUF_LENGTH(buf) == 0 && BUF_MAX(buf) == 0) || (BUF_DATA(buf) != NULL && BUF_MAX(buf) > 0));
	if (BUF_DATA(buf)) {
		free(BUF_DATA(buf));
		BUF_DATA(buf) = NULL;
		BUF_LENGTH(buf) = 0;
		BUF_MAX(buf) = 0;
	}
	
	// if the buffer was created internally, then we need to make it free itself.
	if (buf->internally_created == 1) {
		free(buf);
		return(NULL);
	}
	else {
		return(buf);
	}
}



// add data to the end of our buffer, expanding it if necessary.
void expbuf_add(expbuf_t *buf, const void *data, unsigned int len)
{
	unsigned int avail;
	
	assert(buf);
	assert(data);
	assert(len > 0);
	assert(BUF_LENGTH(buf) <= BUF_MAX(buf));
	assert((BUF_DATA(buf) == NULL && BUF_LENGTH(buf) == 0 && BUF_MAX(buf) == 0) || (BUF_DATA(buf) != NULL && BUF_MAX(buf) > 0));

	avail = BUF_MAX(buf) - BUF_LENGTH(buf);
	if (avail < len) {
		BUF_DATA(buf) = (char*) realloc(BUF_DATA(buf), (BUF_LENGTH(buf) + len));
		assert(BUF_DATA(buf));
		BUF_MAX(buf) = BUF_LENGTH(buf) + len;
		assert(BUF_LENGTH(buf) <= BUF_MAX(buf));
	}

	memmove(BUF_DATA(buf) + BUF_LENGTH(buf), data, len);
	BUF_LENGTH(buf) += len;
	assert(BUF_LENGTH(buf) <= BUF_MAX(buf));
}


// add data to the buffer overwriting what is already there, expanding it if necessary.
void expbuf_set(expbuf_t *buf, const void *data, unsigned int len)
{
	
	assert(buf);
	assert(data);
	assert(len > 0);
	assert(BUF_LENGTH(buf) <= BUF_MAX(buf));
	assert((BUF_DATA(buf) == NULL && BUF_LENGTH(buf) == 0 && BUF_MAX(buf) == 0) || (BUF_DATA(buf) != NULL && BUF_MAX(buf) > 0));

	BUF_LENGTH(buf) = 0;

	if (BUF_MAX(buf) < len) {
		BUF_DATA(buf) = (char*) realloc(BUF_DATA(buf), len);
		assert(BUF_DATA(buf));
		BUF_MAX(buf) = len;
	}

	memmove(BUF_DATA(buf), data, len);
	BUF_LENGTH(buf) = len;
	assert(BUF_LENGTH(buf) <= BUF_MAX(buf));
}


// purge some data from the begining of the buffer (presumably because it has
// been processed).  Moving the remaining data to the beginning of the buffer.
void expbuf_purge(expbuf_t *buf, unsigned int len) {
	assert(buf);
	assert(len > 0);
	assert(BUF_LENGTH(buf) <= BUF_MAX(buf));
	assert((BUF_DATA(buf) == NULL && BUF_LENGTH(buf) == 0 && BUF_MAX(buf) == 0) || (BUF_DATA(buf) != NULL && BUF_MAX(buf) > 0));
	assert(len <= BUF_LENGTH(buf));

	if (len < BUF_LENGTH(buf)) {
		BUF_LENGTH(buf) -= len;
		assert(BUF_LENGTH(buf) > 0);
		memmove(BUF_DATA(buf), BUF_DATA(buf)+len, BUF_LENGTH(buf));
	}
	else {
		assert(BUF_LENGTH(buf) == len);
		BUF_LENGTH(buf) = 0;
	}
}

// The buffer expands as more data is added to it.  This function will shrink
// down, leaving 'extra' padding at the end. 
// this will increase memory usage also if the available space is less than 'extra'.
void expbuf_shrink(expbuf_t *buf, unsigned int extra)
{
	assert(buf);
	assert(BUF_LENGTH(buf) <= BUF_MAX(buf));
	assert((BUF_DATA(buf) == NULL && BUF_LENGTH(buf) == 0 && BUF_MAX(buf) == 0) || (BUF_DATA(buf) != NULL && BUF_MAX(buf) > 0));

	if (extra == 0 && BUF_LENGTH(buf) == 0 && BUF_DATA(buf) != NULL) {
		free(BUF_DATA(buf));
		BUF_DATA(buf) = NULL;
		BUF_MAX(buf) = 0;
	}
	else if ((BUF_MAX(buf) - BUF_LENGTH(buf)) != extra) {
		BUF_DATA(buf) = (char *) realloc(BUF_DATA(buf), BUF_LENGTH(buf)+extra);
		BUF_MAX(buf) = BUF_LENGTH(buf) + extra;
	}
	assert(BUF_LENGTH(buf) <= BUF_MAX(buf));
	assert((BUF_DATA(buf) == NULL && BUF_LENGTH(buf) == 0 && BUF_MAX(buf) == 0) || (BUF_DATA(buf) != NULL && BUF_MAX(buf) > 0));
}


// make sure that the data is null terminated, with enough space, and then return a pointer to the data.  
char * expbuf_string(expbuf_t *buf) {
	assert(buf != NULL);
	assert(BUF_LENGTH(buf) <= BUF_MAX(buf));
	assert((BUF_DATA(buf) == NULL && BUF_LENGTH(buf) == 0 && BUF_MAX(buf) == 0) || (BUF_DATA(buf) != NULL && BUF_MAX(buf) > 0));

	// if there is no space allocated, or not enough space allocated, then we
	// add one char (or create a new one)
	if (BUF_MAX(buf) == BUF_LENGTH(buf) || BUF_DATA(buf) == NULL) {
		BUF_DATA(buf) = (char *) realloc(BUF_DATA(buf), BUF_LENGTH(buf) + 1);
		BUF_MAX(buf)++;
	}
	BUF_DATA(buf)[BUF_LENGTH(buf)] = '\0';
	return(BUF_DATA(buf));
}


void expbuf_print(expbuf_t *buf, const char *format, ...)
{
  va_list ap;
  int redo;
  int n, avail;

	assert(buf);
	assert(format);

	// process the string. Apply directly to the buildbuf.  If buildbuf is not
	// big enough, increase the size and do it again.
	redo = 1;
	while (redo) {
		avail = BUF_MAX(buf)-BUF_LENGTH(buf);
		va_start(ap, format);
		n = vsnprintf(BUF_DATA(buf)+BUF_LENGTH(buf), avail, format, ap);
		va_end(ap);

		assert(n > 0);
		if (n > avail) {
			// there was not enough space, so we need to increase it, and try again.
			expbuf_shrink(buf, n + 1);
			assert(redo);
		}
		else {
			assert(n <= BUF_MAX(buf));
			BUF_LENGTH(buf) += n;
			redo = 0;
		}
	}
}


